closed/src/java.base/share/classes/sun/security/rsa/NativeRSACore.java

```diff
@@ -0,0 +1,144 @@
+import jdk.crypto.jniprovider.NativeCrypto;
+
+import sun.security.jca.JCAUtil;
+
+/**
+ * Core of the RSA implementation. Has code to perform public and private key
+ * RSA operations (with and without CRT for private key ops). Private CRT ops
+ * also support blinding to twart timing attacks.
+ *
+ * The code in this class only does the core RSA operation. Padding and
+ * unpadding must be done externally.
+ *
+ * Note: RSA keys should be at least 512 bits long
+ *
+ * @since   1.5
+ * @author  Andreas Sterbenz
+ */
+public final class NativeRSACore {
+
+    /**
+     * Return the number of bytes required to store the magnitude byte[] of
+     * this BigInteger. Do not count a 0x00 byte toByteArray() would
+     * prefix for 2's complement form.
+     */
+    public static int getByteLength(BigInteger b) {
+        int n = b.bitLength();
+        return (n + 7) >> 3;
+    }
+
+    /**
+     * Perform an RSA public key operation.
+     */
+    public static byte[] rsa(byte[] msg, sun.security.rsa.RSAPublicKeyImpl key)
+        throws BadPaddingException {
+        return crypt_Native(msg, key);
+    }
+
+    /**
+     * Perform an RSA private key operation. Uses CRT if the key is a
+     * CRT key. Set 'verify' to true if this function is used for
+     * generating a signature.
+     */
+    public static byte[] rsa(byte[] msg,sun.security.rsa.RSAPrivateCrtKeyImpl key, boolean verify)
+        throws BadPaddingException {
+        return crtCrypt_Native(msg, key, verify);
+    }
+
+    /**
+     * RSA public key ops. Simple modPow().
+     */
+    synchronized private static byte[] crypt_Native(byte[] msg, sun.security.rsa.RSAPublicKeyImpl key)
+        throws BadPaddingException {
+
+        long nativePtr = key.getNativePtr();
+
+        if (nativePtr == -1) {
+            return null;
+        }
+
+        BigInteger n = key.getModulus();
+        byte[] output = new byte[getByteLength(n)];
+
+        int outputLen = NativeCrypto.RSAEP(msg, msg.length, output, nativePtr);
+
+        if (outputLen == -1) {
+            return null;
+        }
+        return output;
+    }
+
+    /**
+     * RSA private key operations with CRT. Algorithm and variable naming
+     * are taken from PKCS#1 v2.1, section 5.1.2.
+     */
+    synchronized private static byte[] crtCrypt_Native(byte[] msg, sun.security.rsa.RSAPrivateCrtKeyImpl key,
+            boolean verify) throws BadPaddingException {
+        long nativePtr = key.getNativePtr();
+
+        if (nativePtr == -1) {
+            return null;
+        }
+
+        int verifyInt;
+        BigInteger n = key.getModulus();
+        int outputLen = getByteLength(n);
+        byte[] output = new byte[outputLen];
+
+        if(verify) {
+            verifyInt = outputLen;
+        } else {
+            verifyInt = -1;
+        }
+
+        outputLen = NativeCrypto.RSADP(msg, msg.length, output, verifyInt, nativePtr);
+
+        if (outputLen == -1) {
+            return null;
```

Comment:
Should this throw an OutOfMemoryError?

Reply:
@pshipton we have the native crypto setup so any errors in the native code will cause the security provider to switch back to software implementation without any errors to the user. so in the case of any error code outputLen == -1, the NativeRSACore returns null which is checked for in RSACore.java. In RSACore the RSA native implementation is disabled and the software path is executed.

Reply:
ok thanks